scijava/scijava-types/src/main/java/org/scijava/types/extractors/ParameterizedTypeExtractor.java

```diff
@@ -0,0 +1,198 @@
+	 *      ListMap&lt;K, V, T&gt; implements Map&lt;K, V&gt;, List&lt;T&gt;
+	 * </pre>
+	 *
+	 * For such types, we try to fill the type parameters recursively, using
+	 * {@link TypeExtractor} plugins that know how to glean types at runtime from
+	 * specific sorts of objects.
+	 * </p>
+	 * <p>
+	 * For example, {@link IterableTypeExtractor} knows how to guess a {@code T}
+	 * for any {@code Iterable<T>} by examining the type of the elements in its
+	 * iteration. (Of course, this may be inaccurate if the elements in the
+	 * iteration are heterogeneously typed, but for many use cases this guess is
+	 * better than nothing.)
+	 * </p>
+	 * <p>
+	 * In this way, the behavior of the generic type extraction is fully
+	 * extensible, since additional {@link TypeExtractor} plugins can always be
+	 * introduced which extract types more intelligently in cases where more
+	 * <em>a priori</em> knowledge about that type is available at runtime.
+	 * </p>
+	 */
+	public Type reify(final TypeReifier r, final Object o) {
+		if (o == null) return new Any();
+
+		if (o instanceof GenericTyped) {
+			// Object implements the GenericTyped interface; it explicitly declares
+			// the generic type by which it wants to be known. This makes life easy!
+			return ((GenericTyped) o).getType();
+		}
+
+		final Class<?> c = o.getClass();
+		final TypeVariable<?>[] typeVars = c.getTypeParameters();
+		final int numVars = typeVars.length;
+
+		if (numVars == 0) {
+			// if the class is synthetic, we are probably missing something due to
+			// type erasure.
+			if (c.isSynthetic()) {
+				r.log().warn("Object " + o + " is synthetic. " +
+					"Its type parameters are not reifiable and thus will likely cause unintended behavior!");
+			}
+			// Object has no generic parameters; we are done!
+			return c;
+		}
+
+		// Object has parameters which need to be resolved. Let's do it.
+
+		// Here we will store all of our object's resolved type variables.
+		final Map<TypeVariable<?>, Type> resolved = new HashMap<>();
+
+		for (final TypeToken<?> token : TypeToken.of(c).getTypes()) {
+			if (resolved.size() == numVars) break; // Got 'em all!
+
+			final Type type = token.getType();
+			if (!Types.containsTypeVars(type)) {
+				// No type variables are buried in this type; it is useless to us!
+				continue;
+			}
+
+			// Populate relevant type variables from the reified supertype!
+			final Map<TypeVariable<?>, Type> vars = //
+				args(r, o, token.getRawType());
+
+			if (vars != null) {
+				// Remember any resolved type variables.
+				// Note that vars may contain other type variables from other layers
+				// of the generic type hierarchy, which we don't care about here.
+				for (final TypeVariable<?> typeVar : typeVars) {
+					if (vars.containsKey(typeVar)) {
+						resolved.putIfAbsent(typeVar, vars.get(typeVar));
+					}
+				}
+			}
+		}
+
+		// fill in any remaining unresolved type parameters with wildcards
```

Comment:
This comment style differs from earlier comments. Stick with one style.

Reply:
I'm not fully sure what you meant, but I tried to fix it in 035e1c1

Reply:
I meant:
```
// fill in any remaining unresolved type parameters with wildcards
```
vs.
```
// Fill in any remaining unresolved type parameters with wildcards.
```

So, whether to write in complete sentences, or only small case, basically.

For many years I was a small-case no-period guy, but I kind of came around to the idea that reading comments should be like reading anything else written by someone: so make it complete sentences in most situations.

But that's only my personal preference. The main thing is just to be consistent—and I noticed that we had some of the former and others of the latter.

It's a nitpick for sure, though!~